admin/tasks/web.py

```diff
@@ -0,0 +1,240 @@
+        ssl_redir_port=ssl_redir_port,
+        app_host_ip=app_host_ip,
+        extra_config=extra_config,
+    )
+
+    if is_service_started(ctx, "nginx"):
+        reload_service(ctx, "nginx")
+    else:
+        start_service(ctx, "nginx")
+
+
+def setup_nginx(ctx):
+    add_nginx_repository(ctx)
+    install_nginx(ctx)
+    enable_service(ctx, "nginx")
+
+
+@command
+def setup_self_signed_cert(ctx, public_name, ssl_cert, ssl_key):
+    """Installing self-signed TLS certificate"""
+    ctx.sudo(
+        "openssl req -x509 -newkey rsa:4096"
+        f" -keyout {quote(ssl_key)}"
+        f" -out {quote(ssl_cert)}"
+        " -days 365"
+        " -sha256"
+        " -nodes"
+        f" -subj '/C=FR/OU=Zam/CN={public_name}'"
+    )
+
+
+@command
+def setup_letsencrypt_cert(ctx, public_name, email):
+    install_certbot(ctx)
+
+    www_dir = "/var/www/html/letsencrypt"
+
+    with template_local_file(
+        "files/letsencrypt/certbot.ini.template",
+        "files/letsencrypt/certbot.ini",
+        {"host": public_name, "email": email, "www_dir": www_dir},
+    ):
+        sudo_put(ctx, "files/letsencrypt/certbot.ini", "/etc/letsencrypt/certbot.ini")
+
+    with template_local_file(
+        "files/letsencrypt/letsencrypt.conf.template",
+        "files/letsencrypt/letsencrypt.conf",
+        {"www_dir": www_dir},
+    ):
+        if not is_directory(ctx, "/etc/nginx/snippets"):
+            create_directory(ctx, "/etc/nginx/snippets", owner="root")
+
+        sudo_put_if_modified(
+            ctx,
+            "files/letsencrypt/letsencrypt.conf",
+            "/etc/nginx/snippets/letsencrypt.conf",
+        )
+
+    setup_temporary_nginx_config(ctx, public_name, www_dir)
+
+    if is_service_started(ctx, "nginx"):
+        reload_service(ctx, "nginx")
+    else:
+        start_service(ctx, "nginx")
+
+    # Ask for a certificate
+    ctx.sudo(
+        "certbot certonly -c /etc/letsencrypt/certbot.ini --non-interactive --agree-tos"
+    )
+
+    # Renew certificate every week
+    sudo_put(ctx, "files/letsencrypt/ssl-renew", "/etc/cron.weekly/ssl-renew")
+    ctx.sudo("chmod +x /etc/cron.weekly/ssl-renew")
+
+
+@command
+def setup_temporary_nginx_config(ctx, hostname, www_dir):
+    """Setup temporary Nginx config"""
+
+    challenge_dir = posixpath.join(www_dir, ".well-known/acme-challenge")
+    if not is_directory(ctx, challenge_dir):
+        create_directory(ctx, challenge_dir, owner="root")
+
+    with template_local_file(
+        "files/nginx/http.conf.template", "files/nginx/http.conf", {"host": hostname}
+    ):
+        sudo_put(ctx, "files/nginx/http.conf", "/etc/nginx/conf.d/zam.conf")
+
+
+@command
+def setup_final_nginx_config(
+    ctx, hostname, ssl_cert, ssl_key, ssl_redir_port, app_host_ip, extra_config
+):
+
+    if not is_directory(ctx, "/etc/nginx/snippets"):
+        create_directory(ctx, "/etc/nginx/snippets", owner="root")
+
+    sudo_put_if_modified(ctx, "files/nginx/ssl.conf", "/etc/nginx/snippets/ssl.conf")
```

Comment:
That line doesn't look to be effective, to check.

Reply:
Not sure what the problem is. I can see the correct file on the server.

Reply:
See page 48 of the security audit. Even if the file if present, maybe it is not considered?

Reply:
```
→ http --headers https://zam.51.83.111.31.nip.io/
HTTP/1.1 302 Found
Cache-Control: no-store
Connection: keep-alive
Content-Length: 284
Content-Type: text/html; charset=UTF-8
Date: Wed, 18 Dec 2019 23:10:14 GMT
Location: https://zam.51.83.111.31.nip.io/repondeur/identification?source=https%3A%2F%2Fzam.51.83.111.31.nip.io%2Frepondeur%2F
Pragma: no-cache
Server: nginx/1.16.1

→ http --headers https://zam.beta.gouv.fr                                                                                                    2s 164ms
HTTP/1.1 302 Found
Connection: keep-alive
Content-Length: 270
Content-Type: text/html; charset=UTF-8
Date: Wed, 18 Dec 2019 23:10:27 GMT
Location: https://zam.beta.gouv.fr/repondeur/identification?source=https%3A%2F%2Fzam.beta.gouv.fr%2Frepondeur%2F
Server: nginx/1.14.0 (Ubuntu)
Strict-Transport-Security: max-age=15768000; includeSubdomains; preload
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
```